### PR TITLE
metadata reveal for ast-chunking; smart detection of seq length in ollama; auto adjust chunk length for ast to prevent silent truncation

### PR DESCRIPTION
**Pull request by ww2283**
# Enhanced Token Limit Handling with Dynamic Discovery + AST Metadata

This PR primarily adds on top of #152 to extend the metadata revealing when ast chunking is used, also builds on #154 by adding **dynamic token limit discovery** to provide comprehensive token truncation with zero-maintenance for Ollama deployments.

### 1. Dynamic Ollama Token Limit Discovery
- **Zero-maintenance**: Auto-queries Ollama's `/api/show` endpoint for model context limits
- **Future-proof**: Automatically adapts to new models without code changes if ollama is used
- **Fallback safety**: Registry-based limits for LM Studio/OpenAI/unknown models for openai style api 

```python
# Detects limits automatically for Ollama
token_limit = get_model_token_limit("nomic-embed-text", base_url="http://localhost:11434")
# → Queries /api/show, returns actual context_length
```

### 2. Corrected Model Token Limits
- Fixed `nomic-embed-text`: 2048 tokens (was incorrectly 512)
- Verified via `/api/show` endpoint inspection
- Added OpenAI embedding models (text-embedding-3-small/large: 8192)

### 3. AST Metadata Preservation
- Chunking functions now return `List[Dict[str, Any]]` format
- Preserves AST metadata (classes, functions, start_line, end_line)
- Enables metadata display in search results
- Backwards compatible with existing code

### 4. Comprehensive Test Coverage
- **277 lines** of token truncation unit tests
- **553 lines** of AST integration tests
- Tests cover: known/unknown models, single/batch truncation, edge cases
- Current results: **84/89 tests passing** (5 pre-existing RAG environment failures)

## Implementation Details

**Files Changed:**
- `embedding_compute.py`: Dynamic discovery, enhanced truncation, improved logging
- `chunking_utils.py`: Dict-based return format, AST metadata flow
- `cli.py`: AST chunking parameters
- `test_token_truncation.py`: Comprehensive unit tests (new)
- `test_astchunk_integration.py`: AST metadata validation (new)

## Testing

Exclude diskann tests (native C++ threading bugs on macOS):
```bash
uv run pytest -k "not (diskann or test_backend_options or test_large_index)"
```

## Benefits

1. **Zero-maintenance for Ollama**: No hardcoded limits to update
2. **Prevents silent truncation**: Client-side control before server truncates
3. **Better debugging**: Rich metadata for search result provenance
4. **Safer defaults**: Conservative fallbacks (2048) for unknown models
5. **Comprehensive testing**: Verified behavior across scenarios

## Migration Notes

No breaking changes - enhancement is additive. Existing code continues working with static registry, Ollama users automatically benefit from dynamic discovery.

## Checklist

- [X] Tests pass (`uv run pytest`)
- [X] Code formatted (`ruff format` and `ruff check`)
- [X] Pre-commit hooks pass (`pre-commit run --all-files`)

### CODE REVIEW

**Comment by ww2283** [comment]
@ASuresh0524 and @yichuan-w : take a look at this pr and let me know. Btw where's the slack yichuan mentioned to me several days ago? I was not smart enough to find it to join...

**Comment by ASuresh0524** [comment]
hey @ww2283 will look at this by tomorrow night, sorry for the delay

**Comment by yichuan-w** [comment]
Oh the Slack
 link is here
image

File: packages/leann-core/src/leann/chunking_utils.py [hunk 1]
```diff
@@ -5,12 +5,15 @@
...
 from pathlib import Path
-from typing import Optional
+from typing import Any, Optional
 
 from llama_index.core.node_parser import SentenceSplitter
 
 logger = logging.getLogger(__name__)
 
+# Flag to ensure AST token warning only shown once per session
+_ast_token_warning_shown = False
```
Comment (andylizf) [inline packages/leann-core/src/leann/chunking_utils.py:15]: nit: let's use `warnings.filterwarnings`. it's thread-safe.

**Review by andylizf — APPROVED** [review]
LGTM!

**Comment by andylizf** [comment]
Thanks @ww2283  for the awesome PR!

**Comment by ASuresh0524** [comment]
LGTM as well! @yichuan-w @ww2283
